File: .agents/skills/jm-elec-test/scripts/test-website.py
#!/usr/bin/env python3
"""
jm-elec Website Automated Test Suite
Tests all critical paths of the 浙江精铭电子 website.
"""

import urllib.request
import urllib.error
import json
import sys
from datetime import datetime

BASE_URL = "http://localhost:3000"
ADMIN_USER = "admin"
ADMIN_PASS = "jingming2024!"

class TestResult:
    def __init__(self):
        self.passed = []
        self.failed = []
        self.warnings = []

    def ok(self, name, detail=""):
        self.passed.append((name, detail))
        print(f"  ✅ {name}")
        if detail:
            print(f"     {detail}")

    def fail(self, name, detail=""):
        self.failed.append((name, detail))
        print(f"  ❌ {name}")
        if detail:
            print(f"     {detail}")

    def warn(self, name, detail=""):
        self.warnings.append((name, detail))
        print(f"  ⚠️  {name}")
        if detail:
            print(f"     {detail}")

    def summary(self):
        total = len(self.passed) + len(self.failed) + len(self.warnings)
        print("\n" + "="*50)
        print(f"TEST SUMMARY")
        print("="*50)
        print(f"Total:    {total}")
        print(f"Passed:   {len(self.passed)} ✅")
        print(f"Failed:   {len(self.failed)} ❌")
        print(f"Warnings: {len(self.warnings)} ⚠️")
        if self.failed:
            print("\nFailed tests:")
            for name, detail in self.failed:
                print(f"  - {name}: {detail}")
        return len(self.failed) == 0


def http_get(url, cookie=None):
    """Make HTTP GET request and return (status, body, headers)"""
    req = urllib.request.Request(url)
    if cookie:
        req.add_header("Cookie", cookie)
    try:
        with urllib.request.urlopen(req, timeout=10) as resp:
            body = resp.read().decode("utf-8", errors="replace")
            return resp.status, body, dict(resp.headers)
    except urllib.error.HTTPError as e:
        return e.code, e.read().decode("utf-8", errors="replace"), dict(e.headers)
    except Exception as e:
        return 0, str(e), {}


def http_post(url, data=None, json_data=None, cookie=None):
    """Make HTTP POST request"""
    if json_data:
        body = json.dumps(json_data).encode("utf-8")
        req = urllib.request.Request(url, data=body, headers={"Content-Type": "application/json"}, method="POST")
    elif data:
        body = urllib.parse.urlencode(data).encode("utf-8")
        req = urllib.request.Request(url, data=body, headers={"Content-Type": "application/x-www-form-urlencoded"}, method="POST")
    else:
        req = urllib.request.Request(url, method="POST")
    if cookie:
        req.add_header("Cookie", cookie)
    try:
        with urllib.request.urlopen(req, timeout=10) as resp:
            body = resp.read().decode("utf-8", errors="replace")
            return resp.status, body, dict(resp.headers)
    except urllib.error.HTTPError as e:
        return e.code, e.read().decode("utf-8", errors="replace"), dict(e.headers)
    except Exception as e:
        return 0, str(e), {}


def test_frontend_pages(results):
    print("\n📄 FRONTEND PAGES")
    print("-" * 40)

    pages = [
        ("/", "Homepage", ["浙江精铭电子", "我们的优势", "热门产品"]),
        ("/products", "Products Page", ["产品中心", "跨境电商盒装", "快速接线端子"]),
        ("/product/1", "Product Detail", ["产品特点", "立即询盘"]),
        ("/about", "About Page", ["关于我们", "精制造", "发展历程"]),
        ("/contact", "Contact Page", ["联系我们", "在线留言", "提交留言"]),
    ]

    for path, name, keywords in pages:
        status, body, _ = http_get(f"{BASE_URL}{path}")
        if status != 200:
            results.fail(f"{name} ({path})", f"HTTP {status}")
            continue

        missing = [k for k in keywords if k not in body]
        if missing:
            results.fail(f"{name} ({path})", f"Missing keywords: {missing}")
        else:
            results.ok(f"{name} ({path})", f"HTTP 200, all keywords found")

        # Check SEO meta tags
        if "<meta name=\"description\"" not in body and '<meta name="description"' not in body:
            results.warn(f"{name} SEO", "Missing meta description")
        if "<title>" not in body or "</title>" not in body:
            results.warn(f"{name} SEO", "Missing title tag")


def test_static_assets(results):
    print("\n🎨 STATIC ASSETS")
    print("-" * 40)

    assets = [
        "/css/style.css",
        "/css/admin.css",
        "/js/main.js",
        "/images/logo.jpg",
    ]

    for path in assets:
        status, _, _ = http_get(f"{BASE_URL}{path}")
        if status == 200:
            results.ok(f"Asset {path}")
        else:
            results.fail(f"Asset {path}", f"HTTP {status}")


def test_inquiry_submission(results):
    print("\n📨 INQUIRY SUBMISSION")
    print("-" * 40)

    # Test general inquiry
    status, body, _ = http_post(
        f"{BASE_URL}/api/inquiry",
        json_data={
            "name": "Test User",
            "phone": "13800138000",
            "email": "test@example.com",
            "company": "Test Co",
            "message": "This is a test inquiry message",
        }
    )

    if status == 200:
        try:
            data = json.loads(body)
            if data.get("success"):
                results.ok("General inquiry submission", f"ID: {data.get('id')}")
            else:
                results.fail("General inquiry submission", data.get("message"))
        except:
            results.fail("General inquiry submission", "Invalid JSON response")
    else:
        results.fail("General inquiry submission", f"HTTP {status}")

    # Test product inquiry
    status, body, _ = http_post(
        f"{BASE_URL}/api/inquiry",
        json_data={
            "name": "Product Tester",
            "phone": "13900139000",
            "message": "Interested in bulk purchase",
            "product_name": "PCT-222快速接线端子",
            "quantity": "1000只",
        }
    )

    if status == 200:
        try:
            data = json.loads(body)
            if data.get("success"):
                results.ok("Product inquiry submission", f"ID: {data.get('id')}")
            else:
                results.fail("Product inquiry submission", data.get("message"))
        except:
            results.fail("Product inquiry submission", "Invalid JSON response")
    else:
        results.fail("Product inquiry submission", f"HTTP {status}")


def test_admin_login(results):
    print("\n🔐 ADMIN LOGIN")
    print("-" * 40)

    # Test login page
    status, body, _ = http_get(f"{BASE_URL}/admin/login")
    if status == 200 and "精铭电子管理后台" in body:
        results.ok("Admin login page")
    else:
        results.fail("Admin login page", f"HTTP {status}")

    # Test successful login using cookie jar
    import http.cookiejar
    cj = http.cookiejar.CookieJar()
    opener = urllib.request.build_opener(urllib.request.HTTPCookieProcessor(cj))
    
    login_data = urllib.parse.urlencode({"username": ADMIN_USER, "password": ADMIN_PASS}).encode("utf-8")
    req = urllib.request.Request(f"{BASE_URL}/admin/login", data=login_data, headers={"Content-Type": "application/x-www-form-urlencoded"})
    
    try:
        with opener.open(req, timeout=10) as resp:
            final_body = resp.read().decode("utf-8", errors="replace")
            final_url = resp.geturl()
    except urllib.error.HTTPError as e:
        final_body = e.read().decode("utf-8", errors="replace")
        final_url = e.geturl()
    
    # Check if we got session cookie
    session_cookie = None
    for cookie in cj:
        if cookie.name == "connect.sid":
            session_cookie = f"{cookie.name}={cookie.value}"
            break
    
    if session_cookie and "/admin" in final_url:
        results.ok("Admin login success", f"Redirected to {final_url}")
        return session_cookie, opener
    elif session_cookie:
        results.warn("Admin login", f"Got cookie but URL is {final_url}")
        return session_cookie, opener
    else:
        results.fail("Admin login success", "No session cookie received")
        return None, None


def test_admin_dashboard(results, cookie_opener):
    if not cookie_opener:
        results.fail("Admin dashboard", "No login session, skipping")
        return

    print("\n📊 ADMIN DASHBOARD")
    print("-" * 40)

    pages = [
        ("/admin", "Dashboard", ["控制台", "产品总数"]),
        ("/admin/inquiries", "Inquiries", ["询盘管理"]),
        ("/admin/products", "Products", ["产品管理"]),
        ("/admin/uploads", "Uploads", ["图片管理"]),
        ("/admin/settings", "Settings", ["网站设置"]),
    ]

    for path, name, keywords in pages:
        req = urllib.request.Request(f"{BASE_URL}{path}")
        try:
            with cookie_opener.open(req, timeout=10) as resp:
                body = resp.read().decode("utf-8", errors="replace")
                status = resp.status
        except urllib.error.HTTPError as e:
            body = e.read().decode("utf-8", errors="replace")
            status = e.code
        
        if status != 200:
            results.fail(f"Admin {name}", f"HTTP {status}")
            continue

        missing = [k for k in keywords if k not in body]
        if missing:
            results.fail(f"Admin {name}", f"Missing: {missing}")
        else:
            results.ok(f"Admin {name}", f"HTTP 200")

    # Verify inquiries appear in admin
    req = urllib.request.Request(f"{BASE_URL}/admin/inquiries")
    try:
        with cookie_opener.open(req, timeout=10) as resp:
            body = resp.read().decode("utf-8", errors="replace")
    except urllib.error.HTTPError as e:
        body = e.read().decode("utf-8", errors="replace")
    
    if "Test User" in body and "13800138000" in body:
        results.ok("Inquiry data in admin", "General inquiry visible")
    else:
        results.warn("Inquiry data in admin", "General inquiry not found in list")

    if "Product Tester" in body and "PCT-222" in body:
        results.ok("Product inquiry in admin", "Product inquiry visible")
    else:
        results.warn("Product inquiry in admin", "Product inquiry not found in list")


def test_security(results):
    print("\n🔒 SECURITY CHECKS")
    print("-" * 40)

    # Test wrong password
    status, _, _ = http_post(
        f"{BASE_URL}/admin/login",
        data={"username": "admin", "password": "wrongpassword"}
    )
    if status == 200:
        results.ok("Wrong password rejected")
    else:
        results.warn("Wrong password handling", f"HTTP {status}")

    # Test unauthorized admin access
    status, _, _ = http_get(f"{BASE_URL}/admin")
    if status == 302 or status == 301:
        results.ok("Admin requires login", f"Redirected (HTTP {status})")
    else:
        results.warn("Admin access control", f"HTTP {status}, expected redirect")


def test_database_data(results):
    print("\n💾 DATABASE DATA")
    print("-" * 40)

    # Check product count via API (indirectly via product page)
    status, body, _ = http_get(f"{BASE_URL}/products")
    if status == 200:
        # Count product cards
        count = body.count('data-category="')
        if count >= 39:
            results.ok(f"Products loaded", f"Found {count} products")
        else:
            results.warn(f"Products count", f"Only {count} products, expected 39+")

    # Check banners on homepage
    status, body, _ = http_get(f"{BASE_URL}/")
    if status == 200:
        banner_count = body.count('class="carousel-slide')
        if banner_count >= 5:
            results.ok("Banners loaded", f"{banner_count} banners")
        else:
            results.warn("Banners count", f"Only {banner_count} banners")


def main():
    print("="*50)
    print("jm-elec Website Test Suite")
    print(f"Started: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
    print(f"Target:  {BASE_URL}")
    print("="*50)

    results = TestResult()

    # Check if server is running
    status, _, _ = http_get(BASE_URL)
    if status == 0:
        print(f"\n❌ Server not running at {BASE_URL}")
        print("   Please start it first: node server.js")
        sys.exit(1)

    test_frontend_pages(results)
    test_static_assets(results)
    test_inquiry_submission(results)
    cookie, opener = test_admin_login(results)
    test_admin_dashboard(results, opener)
    test_security(results)
    test_database_data(results)

    success = results.summary()

    # Save report
    report_dir = "test-reports"
    import os
    os.makedirs(report_dir, exist_ok=True)
    report_file = os.path.join(report_dir, f"test-report-{datetime.now().strftime('%Y%m%d-%H%M%S')}.txt")

    with open(report_file, "w") as f:
        f.write(f"jm-elec Website Test Report\n")
        f.write(f"Time: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n")
        f.write(f"Target: {BASE_URL}\n")
        f.write("="*50 + "\n\n")
        f.write(f"Passed:   {len(results.passed)}\n")
        f.write(f"Failed:   {len(results.failed)}\n")
        f.write(f"Warnings: {len(results.warnings)}\n\n")
        if results.failed:
            f.write("Failed tests:\n")
            for name, detail in results.failed:
                f.write(f"  - {name}: {detail}\n")

    print(f"\nReport saved to: {report_file}")
    sys.exit(0 if success else 1)


if __name__ == "__main__":
    main()

File: test-website.js

const http = require('http');
const fs = require('fs');
const path = require('path');

// 测试配置
const BASE_URL = 'http://localhost:3000';
const ADMIN_URL = 'http://localhost:3000/admin';
const ADMIN_CREDENTIALS = {
    username: 'admin',
    password: 'jingming2024!'
};

// 测试结果
const testResults = {
    passed: 0,
    failed: 0,
    details: []
};

// 辅助函数：发送HTTP请求
function request(url, method = 'GET', data = null, headers = {}) {
    return new Promise((resolve, reject) => {
        const options = {
            hostname: 'localhost',
            port: 3000,
            path: url.replace(BASE_URL, ''),
            method: method,
            headers: {
                'Content-Type': 'application/json',
                ...headers
            }
        };

        const req = http.request(options, (res) => {
            let body = '';
            res.on('data', (chunk) => {
                body += chunk;
            });
            res.on('end', () => {
                try {
                    const result = {
                        status: res.statusCode,
                        headers: res.headers,
                        body: body,
                        json: body ? JSON.parse(body) : null
                    };
                    resolve(result);
                } catch (e) {
                    resolve({
                        status: res.statusCode,
                        headers: res.headers,
                        body: body,
                        json: null,
                        error: e
                    });
                }
            });
        });

        req.on('error', (e) => {
            reject(e);
        });

        if (data) {
            req.write(JSON.stringify(data));
        }

        req.end();
    });
}

// 辅助函数：断言测试
function assert(condition, message) {
    if (condition) {
        testResults.passed++;
        testResults.details.push(`✓ ${message}`);
        console.log(`✓ ${message}`);
    } else {
        testResults.failed++;
        testResults.details.push(`✗ ${message}`);
        console.log(`✗ ${message}`);
    }
}

// 辅助函数：等待
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

// 测试函数集合
const tests = [
    {
        name: '测试网站首页是否可访问',
        test: async () => {
            const res = await request(BASE_URL);
            assert(res.status === 200, '首页应返回200状态码');
            assert(res.body.includes('浙江精铭电子科技有限公司'), '首页应包含公司名称');
        }
    },
    {
        name: '测试后台登录功能',
        test: async () => {
            // 先获取登录页
            const loginPage = await request(`${ADMIN_URL}/login`);
            assert(loginPage.status === 200, '登录页应可访问');

            // 尝试登录
            const loginRes = await request(`${ADMIN_URL}/login`, 'POST', ADMIN_CREDENTIALS, {
                'Content-Type': 'application/x-www-form-urlencoded'
            });

            // 检查是否登录成功（通过重定向或页面内容）
            assert(loginRes.status === 302 || loginRes.body.includes('dashboard'), '登录应成功并重定向到仪表板');
        }
    },
    {
        name: '测试后台仪表板',
        test: async () => {
            // 先登录
            await request(`${ADMIN_URL}/login`, 'POST', ADMIN_CREDENTIALS, {
                'Content-Type': 'application/x-www-form-urlencoded'
            });

            // 访问仪表板
            const dashboard = await request(`${ADMIN_URL}`);
            assert(dashboard.status === 200, '仪表板应可访问');
            assert(dashboard.body.includes('统计概览'), '仪表板应包含统计信息');
        }
    },
    {
        name: '测试产品管理页面',
        test: async () => {
            // 先登录
            await request(`${ADMIN_URL}/login`, 'POST', ADMIN_CREDENTIALS, {
                'Content-Type': 'application/x-www-form-urlencoded'
            });

            // 访问产品管理页面
            const productsPage = await request(`${ADMIN_URL}/products`);
            assert(productsPage.status === 200, '产品管理页应可访问');
            assert(productsPage.body.includes('产品管理'), '产品管理页应包含标题');
        }
    },
    {
        name: '测试产品编辑功能',
        test: async () => {
            // 先登录
            await request(`${ADMIN_URL}/login`, 'POST', ADMIN_CREDENTIALS, {
                'Content-Type': 'application/x-www-form-urlencoded'
            });

            // 获取产品列表
            const productsPage = await request(`${ADMIN_URL}/products`);
            const productIdMatch = productsPage.body.match(/data-id="(\d+)"/);
            if (productIdMatch) {
                const productId = productIdMatch[1];

                // 获取产品详情
                const productDetail = await request(`${ADMIN_URL}/products/${productId}`);
                assert(productDetail.status === 200, `产品详情页应可访问，ID: ${productId}`);

                // 尝试更新产品
                const updateData = {
                    name: '测试产品更新',
                    category: 'pct',
                    specs: JSON.stringify(['测试规格1', '测试规格2']),
                    description: '这是一个测试产品更新',
                    sort_order: 1,
                    id: productId
                };

                const updateRes = await request(`${ADMIN_URL}/products`, 'POST', updateData, {
                    'Content-Type': 'application/x-www-form-urlencoded',
                    'Cookie': productsPage.headers['set-cookie'] ? productsPage.headers['set-cookie'][0] : ''
                });

                // 检查更新是否成功
                assert(updateRes.status === 302 || updateRes.body.includes('成功'), `产品更新应成功，ID: ${productId}`);

                // 验证更新是否生效
                const updatedProduct = await request(`${ADMIN_URL}/products/${productId}`);
                if (updatedProduct.body) {
                    assert(updatedProduct.body.includes('测试产品更新'), `产品名称应已更新为"测试产品更新"`);
                }
            } else {
                assert(false, '未找到产品ID，无法测试编辑功能');
            }
        }
    },
    {
        name: '测试询盘功能',
        test: async () => {
            // 先登录
            await request(`${ADMIN_URL}/login`, 'POST', ADMIN_CREDENTIALS, {
                'Content-Type': 'application/x-www-form-urlencoded'
            });

            // 访问询盘管理页面
            const inquiriesPage = await request(`${ADMIN_URL}/inquiries`);
            assert(inquiriesPage.status === 200, '询盘管理页应可访问');
            assert(inquiriesPage.body.includes('询盘管理'), '询盘管理页应包含标题');

            // 提交一个测试询盘
            const testInquiry = {
                name: '测试用户',
                phone: '13800138000',
                email: 'test@example.com',
                company: '测试公司',
                product_name: '测试产品',
                quantity: '100',
                message: '这是一个测试询盘'
            };

            const inquiryRes = await request(`${BASE_URL}/api/inquiry`, 'POST', testInquiry);
            assert(inquiryRes.status === 200, '询盘提交应成功');
            assert(inquiryRes.json.success === true, '询盘提交应返回成功状态');

            // 检查询盘是否出现在管理页面
            await sleep(1000); // 等待数据库更新
            const updatedInquiriesPage = await request(`${ADMIN_URL}/inquiries`);
            assert(updatedInquiriesPage.body.includes('测试用户'), '询盘管理页应显示测试询盘');

            // 更新询盘状态
            const inquiryIdMatch = updatedInquiriesPage.body.match(/data-id="(\d+)"/);
            if (inquiryIdMatch) {
                const inquiryId = inquiryIdMatch[1];
                const statusRes = await request(`${ADMIN_URL}/inquiries/${inquiryId}/status`, 'POST', { status: 'done' });
                assert(statusRes.status === 200, '询盘状态更新应成功');
                assert(statusRes.json.success === true, '询盘状态更新应返回成功状态');

                // 删除测试询盘
                const deleteRes = await request(`${ADMIN_URL}/inquiries/${inquiryId}/delete`, 'POST');
                assert(deleteRes.status === 200, '询盘删除应成功');
                assert(deleteRes.json.success === true, '询盘删除应返回成功状态');
            } else {
                assert(false, '未找到询盘ID，无法测试状态更新和删除功能');
            }
        }
    },
    {
        name: '测试联系我们页面',
        test: async () => {
            const contactPage = await request(`${BASE_URL}/contact`);
            assert(contactPage.status === 200, '联系我们页应可访问');
            assert(contactPage.body.includes('在线留言'), '联系我们页应包含在线留言表单');

            // 检查表单提交功能
            const testContact = {
                name: '测试用户',
                phone: '13800138000',
                email: 'test@example.com',
                company: '测试公司',
                product_name: '测试产品',
                quantity: '100',
                message: '这是一个测试留言'
            };

            const contactRes = await request(`${BASE_URL}/api/inquiry`, 'POST', testContact);
            assert(contactRes.status === 200, '联系我们表单提交应成功');
            assert(contactRes.json.success === true, '联系我们表单提交应返回成功状态');
        }
    },
    {
        name: '测试产品详情页',
        test: async () => {
            // 先获取产品列表页
            const productsPage = await request(`${BASE_URL}/products`);
            const productLinkMatch = productsPage.body.match(/href="\/product\/(\d+)"/);

            if (productLinkMatch) {
                const productId = productLinkMatch[1];
                const productDetailPage = await request(`${BASE_URL}/product/${productId}`);
                assert(productDetailPage.status === 200, `产品详情页应可访问，ID: ${productId}`);
                assert(productDetailPage.body.includes('产品详情'), '产品详情页应包含产品详情');
            } else {
                assert(false, '未找到产品链接，无法测试产品详情页');
            }
        }
    },
    {
        name: '测试关于我们页面',
        test: async () => {
            const aboutPage = await request(`${BASE_URL}/about`);
            assert(aboutPage.status === 200, '关于我们页应可访问');
            assert(aboutPage.body.includes('关于我们'), '关于我们页应包含标题');
        }
    }
];

// 主测试函数
async function runTests() {
    console.log('开始测试网站功能...\n');

    for (const test of tests) {
        console.log(`
测试: ${test.name}`);
        try {
            await test.test();
        } catch (error) {
            console.error(`测试出错: ${error.message}`);
            console.error(error.stack);
            testResults.failed++;
            testResults.details.push(`✗ ${test.name} - 错误: ${error.message}`);
        }

        // 添加延迟，避免请求过快
        await sleep(500);
    }

    // 输出测试结果
    console.log('\n\n测试结果:');
    console.log(`通过: ${testResults.passed}`);
    console.log(`失败: ${testResults.failed}`);
    console.log(`总计: ${testResults.passed + testResults.failed}`);

    if (testResults.failed > 0) {
        console.log('\n失败详情:');
        testResults.details.filter(detail => detail.startsWith('✗')).forEach(detail => {
            console.log(detail);
        });
    }

    // 生成测试报告
    const report = `
# 网站功能测试报告

测试时间: ${new Date().toLocaleString()}
测试结果: ${testResults.failed === 0 ? '全部通过' : '存在失败项'}

## 测试统计
- 通过测试: ${testResults.passed}
- 失败测试: ${testResults.failed}
- 总计测试: ${testResults.passed + testResults.failed}

## 测试详情
${testResults.details.map(detail => `- ${detail}`).join('\n')}

## 结论
${testResults.failed === 0 ? '网站功能正常，具备上线条件。' : '网站存在功能问题，需要修复后才能上线。'}
    `;

    // 保存测试报告
    fs.writeFileSync(path.join(__dirname, 'test-report.md'), report);
    console.log('\n测试报告已保存到 test-report.md');
}

// 运行测试
runTests().catch(console.error);
